ccmlib/node.py

```diff
@@ -741,24 +742,26 @@ def wait_for_compactions(self, idle_timeout=300):
         idle_timeout is the time in seconds to wait for progress.
         Total time to wait is undeteremined, as long as we observe forward progress.
         """
-        pending_tasks = None
-        last_change = None
+        prev_task_size, next_task_size = -1, -1
         pattern = re.compile(r"pending tasks:\s*(?P<tasks>\d+)")
-        while not last_change or time.time() - last_change < idle_timeout:
-            output, err = self.nodetool("compactionstats", capture_output=True)
-            m = pattern.search(output)
-            if not m:
+
+        def _wait_for_compactions():
+            nonlocal prev_task_size, next_task_size
+            output = self.nodetool("compactionstats", capture_output=True)[0]
+            match = pattern.search(output)
+            if not match:
                 raise RuntimeError(f"Cannot find 'pending tasks' in nodetool output.\nOutput: {output}")
-            n = int(m.group('tasks'))
-            if n == 0:
-                return
-            if n != pending_tasks:
-                pending_tasks = n
-                last_change = time.time()
-                if n > pending_tasks:     # background progress
-                    self.warning(f"Pending compaction tasks increased from {pending_tasks} to {n} while waiting for compactions.")
-            time.sleep(1)
-        raise TimeoutError(f"Waiting for compactions timed out after {idle_timeout} seconds with {pending_tasks} pending tasks remaining.")
+            next_task_size = int(match.group('tasks'))
+            if next_task_size == 0:
+                return True
+            if next_task_size > prev_task_size:
+                self.warning(f"Pending compaction tasks increased from {prev_task_size} to "
+                             f"{next_task_size} while waiting for compactions.")
+            prev_task_size = next_task_size
+
+        if not wait_for(func=_wait_for_compactions, timeout=idle_timeout, step=1):
```

Comment:
This changed the behavior of `idle_timeout` to be "total timeout".
The idea was that if `next_task_size != 0` after `idle_timeout` but `next_task_size != prev_task_size`, let the loop continue for another `idle_timeout` until either `next_task_size == 0` or there is no progress (`next_task_size == prev_task_size`)

The reason is that we want to support long running compactions (previously there was no timeout at all)

Reply:
I revert my change